### PR TITLE
rubocop: lib/jekyll/document.rb

### PR DESCRIPTION
**Pull request by ayastreb**
#4885

### CODE REVIEW

File: lib/jekyll/document.rb [hunk 1]
```diff
@@ -1,15 +1,16 @@
 # encoding: UTF-8
 
 module Jekyll
+  # rubocop:disable ClassLength
```
Comment (ayastreb) [inline lib/jekyll/document.rb]: I couldn't think of a way to reduce the size of this class (it's 275 lines now, max is 240) 😞 
Any ideas?)

Comment (envygeeks) [inline lib/jekyll/document.rb]: @ayastreb increase the max class size to `300` in `.rubocop.yml` that was an arbitrary number I just randomly set and @parkr decided to keep, I see no problem with raising it.

File: lib/jekyll/document.rb [hunk 2]
```diff
@@ -51,24 +52,28 @@ def data
...
+      if data.key?("date") && !data["date"].is_a?(Time)
+        data["date"] = Utils.parse_date(
+          data["date"].to_s,
           "Document '#{relative_path}' does not have a valid date in the #{source}."
         )
       end
       data
     end
 
+    def merge_categories!(other)
```
Comment (envygeeks) [inline lib/jekyll/document.rb]: `private`, because this belongs to `merge_data!`

File: lib/jekyll/document.rb [hunk 3]
```diff
@@ -51,24 +52,28 @@ def data
...
+      if other.key?("categories") && !other["categories"].nil?
+        if other["categories"].is_a?(String)
+          other["categories"] = other["categories"].split(" ").map(&:strip)
```
Comment (envygeeks) [inline lib/jekyll/document.rb]: Now would be a good time to fix that bad split to `\s+`

File: lib/jekyll/document.rb [hunk 4]
```diff
@@ -81,15 +86,18 @@ def source_file_mtime
...
     end
 
     # The path to the document, relative to the site source.
     #
     # Returns a String path which represents the relative path
     #   from the site source to this document
     def relative_path
-      @relative_path ||= Pathname.new(path).relative_path_from(Pathname.new(site.source)).to_s
+      @relative_path ||= Pathname.new(path)
+        .relative_path_from(Pathname.new(site.source))
```
Comment (envygeeks) [inline lib/jekyll/document.rb]: `Pathutil.new(path).relative_path_from(site.source).to_s` // Note to self.

File: lib/jekyll/document.rb [hunk 5]
```diff
@@ -103,7 +111,7 @@ def output_ext
     #
     # Returns the basename without the file extname.
     def basename_without_ext
-      @basename_without_ext ||= File.basename(path, '.*')
+      @basename_without_ext ||= File.basename(path, ".*")
```
Comment (envygeeks) [inline lib/jekyll/document.rb]: `File.basename(path, File.extname(path))` is much more clear to non-programmers.

File: lib/jekyll/document.rb [hunk 6]
```diff
@@ -156,7 +164,7 @@ def sass_file?
     #
     # Returns true if extname == .coffee, false otherwise.
     def coffeescript_file?
-      '.coffee'.eql?(extname)
+      ".coffee".eql?(extname)
```
Comment (envygeeks) [inline lib/jekyll/document.rb]: `==`

File: lib/jekyll/document.rb [hunk 7]
```diff
@@ -237,7 +245,7 @@ def destination(base_directory)
     def write(dest)
       path = destination(dest)
       FileUtils.mkdir_p(File.dirname(path))
-      File.open(path, 'wb') do |f|
+      File.open(path, "wb") do |f|
```
Comment (envygeeks) [inline lib/jekyll/document.rb]: `File.write(output)` is the same thing without the extra lines.

File: lib/jekyll/document.rb [hunk 8]
```diff
@@ -263,47 +272,61 @@ def read(opts = {})
...
-            raise e
-          end
+        rescue StandardError => e
+          raise e if e.is_a? Jekyll::Errors::FatalException
           Jekyll.logger.error "Error:", "could not read file #{path}: #{e.message}"
         end
       end
     end
 
+    def merge_defaults
```
Comment (envygeeks) [inline lib/jekyll/document.rb]: `private`

File: lib/jekyll/document.rb [hunk 9]
```diff
@@ -263,47 +272,61 @@ def read(opts = {})
...
+          merge_defaults
+          read_content(opts)
           post_read
         rescue SyntaxError => e
           Jekyll.logger.error "Error:", "YAML Exception reading #{path}: #{e.message}"
-        rescue Exception => e
-          if e.is_a? Jekyll::Errors::FatalException
-            raise e
-          end
+        rescue StandardError => e
```
Comment (envygeeks) [inline lib/jekyll/document.rb]: Please revert this and send a separate pull request for this, this has been brought up by me before but there is a larger surface that needs to be tested and thought about, @parkr has his reasons for this and this will block your pull because it's a feature change.

File: lib/jekyll/document.rb [hunk 10]
```diff
@@ -263,47 +272,61 @@ def read(opts = {})
...
 
+    def merge_defaults
+      defaults = @site.frontmatter_defaults.all(
+        relative_path,
+        collection.label.to_sym
+      )
+      merge_data!(defaults, :source => "front matter defaults") unless defaults.empty?
+    end
+
+    def read_content(opts)
```
Comment (envygeeks) [inline lib/jekyll/document.rb]: `private`

File: lib/jekyll/document.rb [hunk 11]
```diff
@@ -263,47 +272,61 @@ def read(opts = {})
...
+    end
+
     def post_read
+      populate_title
+      populate_categories
+      populate_tags
+      generate_excerpt
+    end
+
+    def populate_title
```
Comment (envygeeks) [inline lib/jekyll/document.rb]: `private`

File: lib/jekyll/document.rb [hunk 12]
```diff
@@ -263,47 +272,61 @@ def read(opts = {})
...
-      data['slug'] ||= slug
-      data['ext']  ||= ext
+      data["slug"] ||= slug
+      data["ext"]  ||= ext
+    end
 
-      populate_categories
-      populate_tags
-      generate_excerpt
+    def modify_date(date)
```
Comment (envygeeks) [inline lib/jekyll/document.rb]: `private`

File: lib/jekyll/document.rb [hunk 13]
```diff
@@ -312,16 +335,22 @@ def post_read
     #
     # Returns nothing.
     def categories_from_path(special_dir)
-      superdirs = relative_path.sub(/#{special_dir}(.*)/, '').split(File::SEPARATOR).reject do |c|
+      superdirs = relative_path.sub(/#{special_dir}(.*)/, "")
+        .split(File::SEPARATOR)
+        .reject do |c|
         c.empty? || c.eql?(special_dir) || c.eql?(basename)
```
Comment (envygeeks) [inline lib/jekyll/document.rb]: `==`

File: lib/jekyll/document.rb [hunk 14]
```diff
@@ -394,17 +423,13 @@ def next_doc
       pos = collection.docs.index { |post| post.equal?(self) }
```
Comment (envygeeks) [inline lib/jekyll/document.rb]: `is_a?`

**Comment by envygeeks** [comment]
Thanks! Just a few comments.

File: lib/jekyll/document.rb (hunk 9 above)
Comment (ayastreb) [inline lib/jekyll/document.rb]: Ok, I will revert this and add Rubocop ignore for this

File: lib/jekyll/document.rb (hunk 2 above)
Comment (ayastreb) [inline lib/jekyll/document.rb]: Should I move all the private methods to the bottom of the class below `private`, or do we prefer to leave methods where they are and defined them private, e.g.
```
def merge_data!(other, source: "YAML front matter")
...
end

def merge_categories!(other)
...
end 
private :merge_categories
```
?

Comment (envygeeks) [inline lib/jekyll/document.rb]: At the bottom of the class and each one gets an explicit private.  Rubocop doesn't like that because they consider it useless but I consider it good form to be explicit because one could easily miss a private 20 methods away when quickly viewing a file.

Comment (ayastreb) [inline lib/jekyll/document.rb]: Sorry, when you say each one gets an explicit private, you mean something like this:
```
private
def merge_categories!(other)
...
end

private
def merge_defaults
...
end
```
?
I've checked other files and it seems to be only one `private` keyword per file everywhere.
Just want to make sure I understand you correctly 😄

Comment (envygeeks) [inline lib/jekyll/document.rb]: That is what I mean, see: https://github.com/jekyll/jekyll/blob/master/lib/jekyll/converters/markdown/kramdown_parser.rb, https://github.com/jekyll/jekyll/blob/master/lib/jekyll/commands/serve.rb

File: lib/jekyll/document.rb [hunk 15]
```diff
@@ -237,18 +218,17 @@ def destination(base_directory)
...
     end
 
     # Whether the file is published or not, as indicated in YAML front-matter
     #
-    # Returns true if the 'published' key is specified in the YAML front-matter and not `false`.
+    # Returns true if the 'published' key is
+    # specified in the YAML front-matter and not `false`.
     def published?
-      !(data.key?('published') && data['published'] == false)
+      !(data.key?("published") && data["published"] == false)
```
Comment (pathawks) [inline lib/jekyll/document.rb]: I realize you didn’t change any functionality here, but this line seems to conflict with the description:

> Returns true if the 'published' key is specified in the YAML front-matter and not 'false'.

But it also returns `true` if `published` is _not_ specified.

File: lib/jekyll/document.rb [hunk 16]
```diff
@@ -263,75 +243,19 @@ def read(opts = {})
...
-            merge_data!(data_file, source: "YAML front matter") if data_file
-          end
-
-          post_read
+          merge_defaults
+          read_content(opts)
+          read_data
         rescue SyntaxError => e
           Jekyll.logger.error "Error:", "YAML Exception reading #{path}: #{e.message}"
+        # rubocop: disable RescueException
```
Comment (pathawks) [inline lib/jekyll/document.rb]: Please be sure to enable this again immediately after this block.

File: lib/jekyll/document.rb (hunk 15 above)
Comment (ayastreb) [inline lib/jekyll/document.rb]: Hmm.. should I fix the comment or the method? 😄 
This logic is also duplicated in `Jekyll::Convertible.published?`, it has a different comment there though)

File: lib/jekyll/document.rb (hunk 16 above)
Comment (ayastreb) [inline lib/jekyll/document.rb]: Thx, I forgot to enable it back here 😊

File: lib/jekyll/document.rb (hunk 15 above)
Comment (pathawks) [inline lib/jekyll/document.rb]: The comment, I think.

```ruby
# Returns 'false' if the 'published' key is specified in the YAML front-matter and is 'false'. Otherwise returns 'true'.
```

¯\\\_(ツ)_/¯

**Comment by ayastreb** [comment]
Merged these changes in PR #4961
